aten/src/ATen/native/cuda/zmath.cuh

```diff
@@ -0,0 +1,54 @@
+#pragma once
+
+#include <c10/util/TypeCast.h>
+
+#include <thrust/complex.h>
+
+namespace c10 {
+  // Specialization of is_complex_t() in c10/util/Half.h for thrust::complex types
+  template <typename T>
+  struct is_complex_t<thrust::complex<T>> : public std::true_type {};
+
+  // Specialization of static_cast_with_inter_type in c10/util/TypeCast.h to cast
+  // from std::complex<src_value_t> to thrust::complex<dest_value_t>
+  template <typename dest_value_t, typename src_value_t>
+  struct static_cast_with_inter_type<thrust::complex<dest_value_t>, std::complex<src_value_t>> {
+    C10_HOST_DEVICE static inline thrust::complex<dest_value_t> apply(std::complex<src_value_t> src) {
+      return thrust::complex<dest_value_t>(src.real(), src.imag());
+    }
+  };
+
+  // Specialization of static_cast_with_inter_type in c10/util/TypeCast.h to cast
+  // from thrust::complex<src_value_t> to std::complex<dest_value_t>.
+  // Note Binary compatibility is assumed.
+  template <typename dest_value_t, typename src_value_t>
+  struct static_cast_with_inter_type<std::complex<dest_value_t>, thrust::complex<src_value_t>> {
+    C10_HOST_DEVICE static inline std::complex<dest_value_t> apply(thrust::complex<src_value_t> src) {
+      return reinterpret_cast<std::complex<dest_value_t>&>(src);
```

Comment:
I think this is still UB. Instead of directly reinterpret casting from `thrust::complex` to `std::complex`, we should first construct a temporary array, fill the array with `src.real()` and `src.imag()` and then reinterpret cast the array to `std::complex`

Reply:
I couldn't get the temporary array to work, but as you pointed out the constructor is a constexpr so it should work. I will try again.

Reply:
Ok, I fixed it. Something weird prevented this from working before. Weird.